File: sort.py
arr = [1,8,17,19,2,3,10,12,5,16,7,11,18,20,4,6,9,13,14,15]

# ---------------------------------------
# Selection Sort
# ---------------------------------------
def selection_sort(arr):
    n = len(arr)
    for i in range(n):
        min_index = i
        for j in range(i+1, n):
            if arr[j] < arr[min_index]:
                min_index = j
        arr[i], arr[min_index] = arr[min_index], arr[i]
    return arr

# ---------------------------------------
# Insertion Sort
# ---------------------------------------
def insertion_sort(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        
        while j >= 0 and arr[j] > key:
            arr[j+1] = arr[j]
            j -= 1
        
        arr[j+1] = key

    return arr

# ---------------------------------------
# Bubble Sort
# ---------------------------------------
def bubble_sort(arr):
    n = len(arr)
    for i in range(n):
        for j in range(0, n - i - 1):
            if arr[j] > arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j]
    return arr

# ---------------------------------------
# Quick Sort
# ---------------------------------------
def quick_sort(arr):
    if len(arr) <= 1:
        return arr

    pivot = arr[len(arr)//2]
    left = [x for x in arr if x < pivot]
    mid  = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]

    return quick_sort(left) + mid + quick_sort(right)

# ---------------------------------------
# Merge Sort
# ---------------------------------------
def merge_sort(arr):
    if len(arr) <= 1:
        return arr

    mid = len(arr)//2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    return merge(left, right)


def merge(left, right):
    result = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    result.extend(left[i:])
    result.extend(right[j:])
    return result

# ---------------------------------------
# Heap Sort
# ---------------------------------------
def heapify(arr, n, i):
    largest = i
    left = 2*i + 1
    right = 2*i + 2

    if left < n and arr[left] > arr[largest]:
        largest = left

    if right < n and arr[right] > arr[largest]:
        largest = right

    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)


def heap_sort(arr):
    n = len(arr)

    for i in range(n//2 - 1, -1, -1):
        heapify(arr, n, i)

    for i in range(n-1, -1, -1):
        arr[i], arr[0] = arr[0], arr[i]
        heapify(arr, i, 0)

    return arr

# ---------------------------------------
# Counting Sort
# ---------------------------------------
def counting_sort(arr):
    if len(arr) == 0:
        return arr

    max_val = max(arr)
    count = [0] * (max_val + 1)

    for num in arr:
        count[num] += 1

    result = []
    for i in range(len(count)):
        result.extend([i] * count[i])

    return result

# ---------------------------------------
# Radix Sort
# ---------------------------------------
def counting_sort_for_radix(arr, exp):
    n = len(arr)
    output = [0] * n
    count = [0] * 10

    for number in arr:
        index = (number // exp) % 10
        count[index] += 1

    for i in range(1, 10):
        count[i] += count[i-1]

    for i in range(n-1, -1, -1):
        index = (arr[i] // exp) % 10
        output[count[index] - 1] = arr[i]
        count[index] -= 1

    return output

def radix_sort(arr):
    if len(arr) == 0:
        return arr

    max_val = max(arr)
    exp = 1

    while max_val // exp > 0:
        arr = counting_sort_for_radix(arr, exp)
        exp *= 10

    return arr

# ---------------------------------------
# Shell Sort
# ---------------------------------------
def shell_sort(arr):
    n = len(arr)
    gap = n // 2

    # Reduce the gap each time
    while gap > 0:
        for i in range(gap, n):
            temp = arr[i]
            j = i

            # Insertion sort with gap
            while j >= gap and arr[j - gap] > temp:
                arr[j] = arr[j - gap]
                j -= gap

            arr[j] = temp

        gap //= 2

    return arr


# ---------------------------------------
# Cocktail Shaker Sort
# ---------------------------------------
def cocktail_sort(arr):
    n = len(arr)
    start = 0
    end = n - 1
    swapped = True

    while swapped:
        swapped = False

        # Forward pass
        for i in range(start, end):
            if arr[i] > arr[i + 1]:
                arr[i], arr[i + 1] = arr[i + 1], arr[i]
                swapped = True

        if not swapped:
            break

        swapped = False
        end -= 1

        # Backward pass
        for i in range(end - 1, start - 1, -1):
            if arr[i] > arr[i + 1]:
                arr[i], arr[i + 1] = arr[i + 1], arr[i]
                swapped = True

        start += 1

    return arr



# ---------------------------------------
# Main Program
# ---------------------------------------
while True:
    print("\n--- Sorting Algorithms Menu ---")
    print("1 - Selection Sort")
    print("2 - Insertion Sort")
    print("3 - Bubble Sort")
    print("4 - Quick Sort")
    print("5 - Merge Sort")
    print("6 - Heap Sort")
    print("7 - Counting Sort")
    print("8 - Radix Sort")
    print("0 - Exit")

    choice = int(input("Please choose a sort type: "))

    if choice == 0:
        print("Exiting program...")
        break

    if choice == 1:
        res = selection_sort(arr.copy())
        print("""
Selection Sort:
    We scan the array to find the smallest element and place it at the beginning.
    Then we repeat the process for the rest of the array.
    Complexity:
        Time: O(n²) - Memory: O(1) - Not stable
    Advantage: Simple
    Disadvantage: Very slow for large arrays.
        """)
        print("Result =", res)

    elif choice == 2:
        res = insertion_sort(arr.copy())
        print("""
Insertion Sort:
    We build the sorted list one element at a time by inserting each value
    into its correct position.
    Complexity:
        Time: O(n²) - Memory: O(1) - Stable
    Advantage: Fast if the array is nearly sorted.
    Disadvantage: Slow for large unsorted arrays.
        """)
        print("Result =", res)

    elif choice == 3:
        res = bubble_sort(arr.copy())
        print("""
Bubble Sort:
    We repeatedly compare adjacent elements and swap them if needed.
    Large values "bubble up" to the end.
    Complexity:
        Time: O(n²) - Memory: O(1) - Stable
    Advantage: Very simple to implement.
    Disadvantage: One of the slowest sorting algorithms.
        """)
        print("Result =", res)

    elif choice == 4:
        res = quick_sort(arr.copy())
        print("""
Quick Sort:
    We choose a pivot, split the array into smaller and greater elements,
    and sort the subarrays recursively.
    Complexity:
        Average: O(n log n) - Worst: O(n²) - Memory: O(log n)
    Advantage: Very fast in practice.
    Disadvantage: Bad pivot choice leads to worst-case behavior.
        """)
        print("Result =", res)

    elif choice == 5:
        res = merge_sort(arr.copy())
        print("""
Merge Sort:
    We divide the array into two halves, sort each half, then merge them.
    Complexity:
        Time: O(n log n) - Memory: O(n) - Stable
    Advantage: Very reliable and consistent performance.
    Disadvantage: Uses extra memory.
        """)
        print("Result =", res)

    elif choice == 6:
        res = heap_sort(arr.copy())
        print("""
Heap Sort:
    We build a heap and repeatedly extract the maximum element.
    Complexity:
        Time: O(n log n) - Memory: O(1) - Not stable
    Advantage: Uses constant memory.
    Disadvantage: More complex than other sorts.
        """)
        print("Result =", res)

    elif choice == 7:
        res = counting_sort(arr.copy())
        print("""
Counting Sort:
    We count how many times each value appears.
    Only works for integers in a limited range.
    Complexity:
        Time: O(n + k) - Memory: O(k) - Stable
    Advantage: Extremely fast when range is small.
    Disadvantage: Cannot be used for large or negative ranges.
        """)
        print("Result =", res)

    elif choice == 8:
        res = radix_sort(arr.copy())
        print("""
Radix Sort:
    We sort numbers digit by digit using Counting Sort as a subroutine.
    Complexity:
        Time: O(d * (n + k)) - Memory: O(n + k) - Stable
    Advantage: Very fast for large integers.
    Disadvantage: Works only for integers.
        """)
        print("Result =", res)

    elif choice == 9:
        res = shell_sort(arr.copy())
        print("""
Shell Sort:
    A generalization of insertion sort that allows exchanging elements far apart.
    The algorithm starts with a big gap and reduces it until it becomes 1.
    Complexity:
        Average: O(n^1.3) - Worst: O(n²) - Memory: O(1)
    Advantage: Faster than Insertion Sort for large arrays.
    Disadvantage: Complexity depends on the gap sequence.
        """)
        print("Result =", res)

    elif choice == 10:
        res = cocktail_sort(arr.copy())
        print("""
Cocktail Shaker Sort:
    A variation of Bubble Sort that sorts in both directions.
    This helps move small elements to the beginning faster.
    Complexity:
        Time: O(n²) - Memory: O(1) - Stable
    Advantage: Improves Bubble Sort's performance.
    Disadvantage: Still slow for large arrays.
        """)
        print("Result =", res)

    else:
        print("Invalid choice. Please try again.")

File: sort.c
#include <stdio.h>
#include <stdlib.h>

int arr[] = {1,8,17,19,2,3,10,12,5,16,7,11,18,20,4,6,9,13,14,15};
int n = 20;

/* ---------------------------------------------------
                 SELECTION SORT
----------------------------------------------------*/
void selection_sort(int a[], int n){
    int i, j, min_index, temp;
    for(i = 0; i < n; i++){
        min_index = i;
        for(j = i+1; j < n; j++){
            if(a[j] < a[min_index])
                min_index = j;
        }
        temp = a[i];
        a[i] = a[min_index];
        a[min_index] = temp;
    }
}

/* ---------------------------------------------------
                 INSERTION SORT
----------------------------------------------------*/
void insertion_sort(int a[], int n){
    int i, j, key;
    for(i = 1; i < n; i++){
        key = a[i];
        j = i - 1;
        while(j >= 0 && a[j] > key){
            a[j+1] = a[j];
            j--;
        }
        a[j+1] = key;
    }
}

/* ---------------------------------------------------
                   BUBBLE SORT
----------------------------------------------------*/
void bubble_sort(int a[], int n){
    int i, j, temp;
    for(i = 0; i < n; i++){
        for(j = 0; j < n-i-1; j++){
            if(a[j] > a[j+1]){
                temp = a[j];
                a[j] = a[j+1];
                a[j+1] = temp;
            }
        }
    }
}

/* ---------------------------------------------------
                      QUICK SORT
----------------------------------------------------*/
void quick_sort(int a[], int low, int high){
    if(low >= high) return;

    int pivot = a[(low + high) / 2];
    int i = low, j = high, temp;

    while(i <= j){
        while(a[i] < pivot) i++;
        while(a[j] > pivot) j--;

        if(i <= j){
            temp = a[i];
            a[i] = a[j];
            a[j] = temp;
            i++; j--;
        }
    }

    quick_sort(a, low, j);
    quick_sort(a, i, high);
}

/* ---------------------------------------------------
                      MERGE SORT
----------------------------------------------------*/
void merge(int a[], int l, int m, int r){
    int i = l, j = m+1, k = 0;
    int size = r - l + 1;
    int* temp = malloc(size * sizeof(int));

    while(i <= m && j <= r){
        if(a[i] <= a[j]) temp[k++] = a[i++];
        else temp[k++] = a[j++];
    }

    while(i <= m) temp[k++] = a[i++];
    while(j <= r) temp[k++] = a[j++];

    for(i = l, k = 0; i <= r; i++, k++)
        a[i] = temp[k];

    free(temp);
}

void merge_sort(int a[], int l, int r){
    if(l >= r) return;
    int m = (l + r) / 2;
    merge_sort(a, l, m);
    merge_sort(a, m+1, r);
    merge(a, l, m, r);
}

/* ---------------------------------------------------
                       HEAP SORT
----------------------------------------------------*/
void heapify(int a[], int n, int i){
    int largest = i;
    int left = 2*i + 1;
    int right = 2*i + 2;

    if(left < n && a[left] > a[largest])
        largest = left;

    if(right < n && a[right] > a[largest])
        largest = right;

    if(largest != i){
        int temp = a[i];
        a[i] = a[largest];
        a[largest] = temp;
        heapify(a, n, largest);
    }
}

void heap_sort(int a[], int n){
    int i, temp;
    for(i = n/2 - 1; i >= 0; i--)
        heapify(a, n, i);

    for(i = n - 1; i >= 0; i--){
        temp = a[0];
        a[0] = a[i];
        a[i] = temp;
        heapify(a, i, 0);
    }
}

/* ---------------------------------------------------
                     COUNTING SORT
----------------------------------------------------*/
void counting_sort(int a[], int n){
    int max = a[0], i;
    for(i = 1; i < n; i++)
        if(a[i] > max) max = a[i];

    int* count = calloc(max+1, sizeof(int));
    int* out = malloc(n * sizeof(int));

    for(i = 0; i < n; i++)
        count[a[i]]++;

    int idx = 0;
    for(i = 0; i <= max; i++)
        while(count[i]--)
            out[idx++] = i;

    for(i = 0; i < n; i++)
        a[i] = out[i];

    free(count);
    free(out);
}

/* ---------------------------------------------------
                       RADIX SORT
----------------------------------------------------*/
void counting_sort_radix(int a[], int n, int exp){
    int out[n], count[10] = {0};
    int i;

    for(i = 0; i < n; i++)
        count[(a[i] / exp) % 10]++;

    for(i = 1; i < 10; i++)
        count[i] += count[i-1];

    for(i = n-1; i >= 0; i--){
        int digit = (a[i] / exp) % 10;
        out[count[digit] - 1] = a[i];
        count[digit]--;
    }

    for(i = 0; i < n; i++)
        a[i] = out[i];
}

void radix_sort(int a[], int n){
    int max = a[0], i;
    for(i = 1; i < n; i++)
        if(a[i] > max) max = a[i];

    for(int exp = 1; max / exp > 0; exp *= 10)
        counting_sort_radix(a, n, exp);
}

/* ---------------------------------------------------
              SHELL SORT
----------------------------------------------------*/
void shell_sort(int a[], int n){
    for(int gap = n/2; gap > 0; gap /= 2){
        for(int i = gap; i < n; i++){
            int temp = a[i], j;
            for(j = i; j >= gap && a[j-gap] > temp; j -= gap)
                a[j] = a[j-gap];
            a[j] = temp;
        }
    }
}

/* ---------------------------------------------------
            COCKTAIL SHAKER SORT
----------------------------------------------------*/
void cocktail_sort(int a[], int n){
    int start = 0, end = n - 1, swapped = 1;

    while(swapped){
        swapped = 0;

        for(int i = start; i < end; i++){
            if(a[i] > a[i+1]){
                int t = a[i]; a[i] = a[i+1]; a[i+1] = t;
                swapped = 1;
            }
        }

        if(!swapped) break;
        swapped = 0;
        end--;

        for(int i = end - 1; i >= start; i--){
            if(a[i] > a[i+1]){
                int t = a[i]; a[i] = a[i+1]; a[i+1] = t;
                swapped = 1;
            }
        }
        start++;
    }
}

/* ---------------------------------------------------
                  UTILS
----------------------------------------------------*/
void copy_array(int dst[], int src[]){
    for(int i = 0; i < n; i++) dst[i] = src[i];
}

void print_array(int a[]){
    for(int i = 0; i < n; i++)
        printf("%d ", a[i]);
    printf("\n");
}

/* ---------------------------------------------------
                      MENU
----------------------------------------------------*/
int main(){
    int choice;
    int a[20];

    while(1){
        printf("\n--- Sorting Algorithms Menu ---\n");
        printf("1 - Selection Sort\n");
        printf("2 - Insertion Sort\n");
        printf("3 - Bubble Sort\n");
        printf("4 - Quick Sort\n");
        printf("5 - Merge Sort\n");
        printf("6 - Heap Sort\n");
        printf("7 - Counting Sort\n");
        printf("8 - Radix Sort\n");
        printf("9 - Shell Sort\n");
        printf("10 - Cocktail Sort\n");
        printf("0 - Exit\n");
        printf("Enter choice: ");
        scanf("%d", &choice);

        if(choice == 0){
            printf("Exiting...\n");
            break;
        }

        copy_array(a, arr);

        switch(choice){

            case 1: selection_sort(a, n); printf("Selection Sort Result:\n"); break;
            case 2: insertion_sort(a, n); printf("Insertion Sort Result:\n"); break;
            case 3: bubble_sort(a, n); printf("Bubble Sort Result:\n"); break;
            case 4: quick_sort(a, 0, n-1); printf("Quick Sort Result:\n"); break;
            case 5: merge_sort(a, 0, n-1); printf("Merge Sort Result:\n"); break;
            case 6: heap_sort(a, n); printf("Heap Sort Result:\n"); break;
            case 7: counting_sort(a, n); printf("Counting Sort Result:\n"); break;
            case 8: radix_sort(a, n); printf("Radix Sort Result:\n"); break;
            case 9: shell_sort(a, n); printf("Shell Sort Result:\n"); break;
            case 10: cocktail_sort(a, n); printf("Cocktail Sort Result:\n"); break;

            default: printf("Invalid choice.\n"); continue;
        }

        print_array(a);
    }
}
